riot_pal/dut_shell.py

```diff
@@ -57,18 +66,82 @@ def send_cmd(self, send_cmd):
             if self.SUCCESS in response:
                 clean_msg = response.replace(self.SUCCESS, '')
                 cmd_info['msg'] = clean_msg.replace('\n', '')
-                cmd_info['result'] = self.RESULT_SUCCESS
+                cmd_info['result'] = RESULT_SUCCESS
                 cmd_info['data'] = self._try_parse_data(cmd_info['msg'])
                 break
 
             if self.ERROR in response:
                 clean_msg = response.replace(self.ERROR, '')
                 cmd_info['msg'] = clean_msg.replace('\n', '')
-                cmd_info['result'] = self.RESULT_ERROR
+                cmd_info['result'] = RESULT_ERROR
                 break
-            response = self._readline()
+            response = self.dev._readline()
 
         if response == '':
-            cmd_info['result'] = self.RESULT_TIMEOUT
-            logging.debug(self.RESULT_TIMEOUT)
+            cmd_info['result'] = RESULT_TIMEOUT
+            logging.debug(RESULT_TIMEOUT)
+        return cmd_info
+
+
+class JSONParser:
+    """Handles parsing of specific json data
+
+    Args:
+        dev -> device to connect send and recieve data
+    """
+
+    def __init__(self, dev):
+        self.dev = dev
+
+    def _send_cmd(self, cmd_to_send, end_key='result'):
+        self.dev._write(cmd_to_send)
+        cmd_info = {}
+        while end_key not in cmd_info:
+            line = ""
+            try:
+                line = self.dev._readline()
+                cmd_info.update(json.loads(line))
+            except json.decoder.JSONDecodeError:
+                if 'msg' not in cmd_info:
+                    cmd_info['msg'] = []
+                cmd_info['msg'].append(line)
+            except TimeoutError:
+                cmd_info['result'] = RESULT_TIMEOUT
         return cmd_info
+
+    def send_and_parse_cmd(self, cmd_to_send):
+        """Returns a dictionary with information from the event
+        Args:
+            send_cmd(str): The command to write to the device
+            to_byte_array: If True and data is bytes leave it as an array
+            timeout: Optional timeout value for command specific timeouts
+        Returns:
+            dict:
+            The return hold dict values in the following keys::
+            msg - The message from the response, only used for information.
+            cmd - The command sent, used to track what has occured.
+            data - Parsed information of the data requested.
+            result - Either success, error or timeout.
+        """
+
+        cmd_info = {'cmd': cmd_to_send}
+        cmd_info.update(self._send_cmd(cmd_to_send))
+        return cmd_info
+
+
+class DutShell:
```

Comment:
run this through flake8 or a linter, it will complain about missing doc strings 😉

Reply:
The most important, documentation!